ptb2musicxml.1
.TH ptb2musicxml 1 "4 May 2004"
.SH NAME
ptb2musicxml \- PowerTab file to ASCII file converter
.SH SYNOPSIS
.PP
.B ptb2musicxml 
[-d]
[-o \fIoutput-file\fP]
\fIpowertab-file.ptb\fP
.RI
.SH DESCRIPTION
\fBptb2ascii\fP is a program that takes a file generated by the PowerTab 
Tablature editor and generates an MusicXML file containing the same data.

.PP
.SH OPTIONS
.PP
.IP "--help"
Show all available options.
.IP "-d"
Run in debug mode. This will generate a lot of output to stderr.
.IP "-o \fIoutput-file\fP"
Filename of the ASCII file that should be generated. If this is not 
specified, the ASCII file will be written to stdout.
.SH "SEE ALSO"
.BR http://jelmer.vernstok.nl/
.PP
.BR http://www.power-tab.net/
.PP
.BR http://www.musicxml.org/

.SH BUGS
.PP
Please report any bugs to Jelmer Vernooij at \fBjelmer@samba.org\fP.
.SH LICENSE
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful, but
\fBWITHOUT ANY WARRANTY\fR; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU 
General Public License for more details.
.PP
You should have received a copy of the GNU General Public License 
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple PLace, Suite 330, Boston, MA  02111-1307  USA
.SH AUTHOR
.BR
 Jelmer Vernooij <jelmer@samba.org>


NEWS
All changes, unless specified otherwise, made by Jelmer Vernooij
<jelmer@samba.org>.

0.5.1	UNRELEASED

 CHANGES
  
  * Change license to GPLv3 or later.

0.5.0:
 * Portability improvements. 
 * Switched VCS to bazaar.
 * Support a few more fields (based on ptparse released by Brad).
 * Initial testsuite added.
 * Use soname.

0.4.3:
 * Fix MacOSX shared library build
 * Add -t option to ptbinfo
 * Saner internals in ptb2ly
 * Support different tunings

0.4.2:
 * Add DTD for XML generated by ptb2xml
 * Add ptb2abc utility
 * Default to always printing errors
 * Allow overriding or disabling of printing function used for errors
 	and warnings
 * Support for some more types of data
 * Fix bugs when reading odd combinations of symbols (multiple 
	 coda symbols, etc)
 * Add ptbdict utility
 * Add parser for ptb tuning dictionaries (thanks to Matthias 
		Vogelgesang for information about the file format)
 * Move variables from Makefile to Makefile.settings.in
 * Support cross-compiling
 * Add gp2ly utility
 * Add ptb_free function
 * Fix bug in reading bends for ptb 
 * Correctly parse CMusicBar (thanks to Matthias Vogelgesang)
 * pragma(lib) is MSVC-specific (thanks to Matthias Vogelgesang)

0.4.1:
 * Remove dependency on glib
 * Support \book in lilypond output
 * (ptb2ly) triplet support
 * Write support (ptabtools can read ptb files it writes out, I'm not sure
				  about powertab itself)
 * Start working on GuitarPro support
 * Move to autoconf
 * Win32 build works now
 * Add setup files for Win32 build
 * Small fixes
 * Default to <basename>.{txt,ly,xml} as output file and support -o - for
 writing to stdout.
 * More diagnostic messages
 * (ptb2xml) Escape ampersands in output
 * (ptb2xml) Add -f option
 * (ptb2ly) Include version information
 * (ptb2ly) Generate fewer lilypond output files, make midi work

0.3:
 * Put chords at the right place
 * Support dots
 * Keep beaming information
 * Escape invalid lilypond stuff
 * Generate MusicXML using XSL script
 * Fix a couple of problems with notes in lower strings
 * Various other improvements

0.2:
 * Better parsing of the structure
 * Add ptb2xml
 * Add ptb2musicxml
 * Add ptbinfo
 * Add support for some more properties
 * More asserts
 * Add chord name support
 * Don't append length of note if not necessary in ptb2ly
 * Fix chord handling

0.1:
 * Initial release


ptb2musicxml.c
/*
	(c) 2004: Jelmer Vernooij <jelmer@samba.org>

	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 2 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <stdio.h>
#include <errno.h>
#include <popt.h>
#include <libxml/xmlmemory.h>
#include <libxml/parser.h>
#include "ptb.h"

int main(int argc, const char **argv) 
{
	struct ptbf *ret;
	int debugging = 0;
	xmlNodePtr root_node;
	xmlDocPtr doc;
	xmlDtdPtr dtd;
	int c;
	int version = 0;
	char *output = NULL;
	poptContext pc;
	struct poptOption options[] = {
		POPT_AUTOHELP
		{"debug", 'd', POPT_ARG_NONE, &debugging, 0, "Turn on debugging output" },
		{"outputfile", 'o', POPT_ARG_STRING, &output, 0, "Write to specified file", "FILE" },
		{"version", 'v', POPT_ARG_NONE, &version, 'v', "Show version information" },
		POPT_TABLEEND
	};

	pc = poptGetContext(argv[0], argc, argv, options, 0);
	poptSetOtherOptionHelp(pc, "file.ptb");
	while((c = poptGetNextOpt(pc)) >= 0) {
		switch(c) {
		case 'v':
			printf("ptb2ascii Version "PTB_VERSION"\n");
			printf("(C) 2004 Jelmer Vernooij <jelmer@samba.org>\n");
			exit(0);
			break;
		}
	}
			
	ptb_set_debug(debugging);
	
	if(!poptPeekArg(pc)) {
		poptPrintUsage(pc, stderr, 0);
		return -1;
	}
	ret = ptb_read_file(poptGetArg(pc));
	
	if(!ret) {
		perror("Read error: ");
		return -1;
	} 

	doc = xmlNewDoc(BAD_CAST "1.0");
	root_node = xmlNewNode(NULL, BAD_CAST "score-partwise");
	xmlDocSetRootElement(doc, root_node);

	dtd = xmlCreateIntSubset(doc, BAD_CAST "root", BAD_CAST "-//Recordare//DTD MusicXML 1.0 Partwise//EN", BAD_CAST "http://www.musicxml.org/dtds/partwise.dtd");
		
	xmlSaveFormatFileEnc(output?output:"-", doc, "UTF-8", 1);

	xmlFreeDoc(doc);

	xmlCleanupParser();

	return 0;
}
